### PR TITLE
update light client state prover

### PR DESCRIPTION
**Pull request by alxiong**
### This PR:

updates the `hotshot-state-prover` with the new split between the static `StakeTableState` and the dynamic `LightClientState`

Currently there are still errors in my code.

### CODE REVIEW

File: hotshot-state-prover/src/circuit.rs [hunk 1]
```diff
@@ -476,9 +448,8 @@ mod tests {
             .is_err());
 
         // bad path: total weight doesn't meet the threshold
-        // bit vector with total weight 23
         let bad_bit_vec = [
-            true, true, true, true, true, false, false, true, false, false,
+            false, false, true, false, true, false, false, true, false, false,
```
Comment (alysiahuggins) [inline hotshot-state-prover/src/circuit.rs:452]: can you explain this change a bit?

File: contracts/src/libraries/PlonkVerifier.sol [hunk 2]
```diff
@@ -262,31 +261,30 @@ library PlonkVerifier {
...
             let wire1Ptr := mload(add(proof, 0x20))
-            mstore(add(dataPtr, 0x6c0), mload(wire1Ptr)) // wire1.x
-            mstore(add(dataPtr, 0x6e0), mload(add(wire1Ptr, 0x20))) // wire1.y
+            mstore(add(dataPtr, 0x6a0), mload(wire1Ptr)) // wire1.x
+            mstore(add(dataPtr, 0x6c0), mload(add(wire1Ptr, 0x20))) // wire1.y
             let wire2Ptr := mload(add(proof, 0x40))
-            mstore(add(dataPtr, 0x700), mload(wire2Ptr)) // wire2.x
-            mstore(add(dataPtr, 0x720), mload(add(wire2Ptr, 0x20))) // wire2.y
+            mstore(add(dataPtr, 0x6e0), mload(wire2Ptr)) // wire2.x
+            mstore(add(dataPtr, 0x700), mload(add(wire2Ptr, 0x20))) // wire2.y
```
Comment (alysiahuggins) [inline contracts/src/libraries/PlonkVerifier.sol:274]: can you also explain this value change?

File: hotshot-state-prover/src/circuit.rs [hunk 3]
```diff
@@ -335,7 +301,10 @@ where
     circuit.enforce_true(sig_ver_result.0)?;
 
     circuit.finalize_for_arithmetization()?;
-    Ok((circuit, public_inputs.into()))
+    Ok((
+        circuit,
+        GenericPublicInput::new(lightclient_state.clone(), stake_table_state.clone()),
```
Comment (alysiahuggins) [inline hotshot-state-prover/src/circuit.rs:306]: yes I was thinking of doing something like this 👏

File: hotshot-state-prover/src/mock_ledger.rs [hunk 4]
```diff
@@ -201,25 +188,26 @@ impl MockLedger {
...
+    //     }
 
-        self.sync_stake_table(num_reg, num_exit);
-    }
+    //     self.sync_stake_table(num_reg, num_exit);
+    // }
 
     /// Return the light client state and proof of consensus on this finalized state
-    pub fn gen_state_proof(&mut self) -> (GenericPublicInput<F>, Proof, GenericStakeTableState<F>) {
+    pub fn gen_state_proof(&mut self) -> (GenericPublicInput<F>, Proof) {
```
Comment (alysiahuggins) [inline hotshot-state-prover/src/mock_ledger.rs:210]: love this

File: hotshot-state-prover/src/service.rs [hunk 5]
```diff
@@ -590,11 +570,11 @@ mod test {
         Vec<(StateSignKey, StateVerKey)>,
         StakeTable<BLSPubKey, StateVerKey, CircuitField>,
     ) {
-        let pp = MockSystemParam::init(BLOCKS_PER_EPOCH);
+        let pp = MockSystemParam::init();
```
Comment (alysiahuggins) [inline hotshot-state-prover/src/service.rs:573]: 👍

File: hotshot-state-prover/src/circuit.rs (hunk 1 above)
Comment (alxiong) [reply hotshot-state-prover/src/circuit.rs:452]: ah, the making sure the accumulated weight of the signers is lower than threshold. previously we set a higher threshold, now it's computed as 1/3 +1, so I let fewer nodes to vote.

File: contracts/src/libraries/PlonkVerifier.sol (hunk 2 above)
Comment (alxiong) [reply contracts/src/libraries/PlonkVerifier.sol:274]: yes, so we removed a public input field (from 8 to 7), 
so all these values, which were laying out consecutively in the memory space, are moved up in their memory location by `0x20` (one field element is 32 bytes)

**Comment by alxiong** [comment]
This marks the closure of the #1939.

Please "merge with commit" instead of squash merge here, lots of useful history.

**Comment by alxiong** [comment]
@mrain can you review any changes under `hotshot-state-prover`? 
thanks!

**Comment by alxiong** [comment]
ah, and I don't quite understand the gas increase for the PlonkVerifier, but we can investigate in another PR. 
I'll look at it over the weekend

cc @philippecamacho

**Review by mrain — APPROVED** [review]
state prover part LGTM

**Review by philippecamacho — APPROVED** [review]
LGTM

**Pull request merged**